### ISSUE TITLE
CU - Research the potential of a snackbar silent failure

### ISSUE THREAD

**Issue by dumathane** [department-of-veterans-affairs/va-mobile-app#9618]
## Proposed Change

Research our current snackbar implementation and confirm all snackbars currently wait on API communication to complete?
If any snackbars do not wait for an api to complete, can we fail to send a snackbar on timeout or excessive processing time? 
Can we trigger an api return at nearly the same time as the 'back' button on a given screen and erase the snackbar before it finishes rendering?
Can we tinker with our snackbar dismissal code to ignore snackbar dismissals from navigation if the snackbar itself is a failure message?

## Coding Time Estimation

3
	
## Checklist
- [ ] Add the relevant team label (Health, global, design system, API, Qa and Release etc.)
- [ ] Attach to ticket to the relevant Team Tech Debt Epic epic (old frontend engineering epic is no longer in use as each team is managing their code's technical debt and code upkeep work)

**Comment by TKDickson** [comment]
I remembered the existence of #5964 today, which used to document a silent failure in editing address in letters, under specific circumstances. We've significantly re-architected the app (BE and FE) since that point, and the staging user I have on the ticket no longer has the same authorized services in staging (`lettersAndDocuments true, userProfileUpdate false`) to be able to determine the 'true' behavior in the app (key here: which of the many contact-info-related API calls would fail? what would they return in their fails? how would the app UI react to that now?).

**Comment by TKDickson** [comment]
Also @alexandec - here are the silent failure errors we've fixed previously in the app (that I know of), and how we've fixed them. Maybe it'll help to identify patterns/parallel things to be scanning for in the audit?

- [Silent failure saving email address 2400](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/2400), [fix PR](https://github.com/department-of-veterans-affairs/va-mobile-app/pull/4437) 
- [Silent failure saving direct deposit info (this one I'm not sure if it was ever released, or caught pre-release) 1588](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/1588), [fix PR](https://github.com/department-of-veterans-affairs/va-mobile-app/pull/1636)
- [Silent failure saving preferred name (this one I know did not ever reach production) 4966](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/4966), [fix PR](https://github.com/department-of-veterans-affairs/va-mobile-app/pull/5425)
- [Silent failure changing notification preferences 5036](https://app.zenhub.com/workspaces/va-mobile-60f1a34998bc75000f2a489f/issues/gh/department-of-veterans-affairs/va-mobile-app/5036), [fix PR](https://github.com/department-of-veterans-affairs/va-mobile-app/pull/6889)

**Comment by alexandec** [comment]
### Research our current snackbar implementation and confirm all snackbars currently wait on API communication to complete?

We show loading/saving screens while the API communication is in progress. I checked all the instances of `showSnackBar` and those are accompanied by display of a loading screen. The snackbar is only shown when the request completes with either success or failure. This prevents the issue of the user navigating to a different screen prior to the snackbar being shown.

The loading screen has a Cancel button on it, but no other buttons. The user can't navigate to another page while the loading screen is shown and the request is still in progress, other than by pressing the Cancel button. If they press the Cancel button, then navigate to another page while the request is still in progress, the snackbar is not shown when the request completes. The onSuccess/onError handlers are not run in that case, so the code to show the snackbar doesn't run. I suppose this could be considered a silent error in a sense, but since the user clicked Cancel, their expectation is probably that the request didn't complete.

### If any snackbars do not wait for an api to complete, can we fail to send a snackbar on timeout or excessive processing time?

I didn't find any snackbars that were shown before the request completes. Almost all our calls to `showSnackBar` are placed inside onSuccess and onError callbacks which don't run until the request is done. I tested throttling the request with Charles, to slow the request down. If the request takes a long time to complete, the loading screen stays up for a long time. The snackbar isn't shown until the loading screen goes away.

### Can we trigger an api return at nearly the same time as the 'back' button on a given screen and erase the snackbar before it finishes rendering?

In my experimentation, the snackbar always finished rendering before I could click the Back button. But I can definitely make the snackbar disappear quickly before it would be possible to fully read it, by pressing the Back button as soon as possible when it appears. So there's some risk that a user would do so.

### Can we tinker with our snackbar dismissal code to ignore snackbar dismissals from navigation if the snackbar itself is a failure message?

I investigated the internal state of the snackBar object and found a way to determine whether it's an error snackbar. I made use of it in `CloseSnackbarOnNavigation` as follows:

```
const isErrorSnackbar = snackBar?.state?.toasts[0]?.data?.isError // undocumented internal state
if (!isErrorSnackbar) {
  snackBar?.hideAll()
}
```

I don't recommend using this method in production because the internal state is not documented and could change at any time. However, it does work. Success snackbars are closed on navigation as before. Error snackbars stay open despite navigation through multiple screens. To actually use this approach, I would recommend we store the snackbar state elsewhere, such as in the Redux store.

There's another issue with this method however. It occurs when the error snackbar has a "Try again" (retry) button. If the error snackbar appears, then the user navigates to a different screen, then hits "Try again", the retry does not execute. Retry works fine as long as the user stays on the same screen, but otherwise not. The root cause is that the onSuccess/onError callbacks don't run after navigation.

### Why don't the onSuccess/onError callbacks run after the user navigates to a different screen?

React Query allows success and failure callbacks to be passed either to `useMutation`, to `mutate`, or to both. Our onSuccess and onError callbacks don't run after navigation because 1) we pass them as arguments to `mutate`, and 2) the component where they were created was unmounted when the user left the screen. From the [TanStack Query docs](https://tanstack.com/query/latest/docs/framework/react/guides/mutations?from=reactQueryV3#mutation-side-effects):

> You might find that you want to trigger additional callbacks beyond the ones defined on useMutation when calling mutate. This can be used to trigger component-specific side effects. To do that, you can provide any of the same callback options to the mutate function after your mutation variable. Supported options include: onSuccess, onError and onSettled. Please keep in mind that those additional callbacks won't run if your component unmounts before the mutation finishes.

This behavior is intentional according to React Query's maintainer, [who recommends](https://tkdodo.eu/blog/mastering-mutations-in-react-query#some-callbacks-might-not-fire):

> Do things that are absolutely necessary and logic related (like query invalidation) in the useMutation callbacks. Do UI related things like redirects or showing toast notifications in mutate callbacks. If the user navigated away from the current screen before the mutation finished, those will purposefully not fire.

We are currently following React Query's recommended practices in showing the toast notification (snackbar) only on the screen where it was created. But if we want to allow navigation to a different screen before the request completes, and still show the snackbar, we'll need to make a change.

### How can we fix this callback issue?

To make the success and failure callbacks work after navigating, we need to pass them into `useMutation` instead of `mutate`. In my experimental example I added parameters for the callbacks like:

```
export const useSavePhoneNumber = (onSuccess: () => void, onError: (error: unknown) => void) => { // added these parameters
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: savePhoneNumber,
    onSuccess: () => {
      ...
      onSuccess() // added this
    },
    onError: (error) => {
      ...
      onError(error) // added this
    },
  })
}
```

Then I modified EditPhoneNumberScreen as needed to pass the callbacks when it calls the hook. I had to move a bunch of code higher in the component so everything is defined before passing it into the hook. The result is that onSuccess and onError always run when the request completes, whether the user stays on the EditPhoneNumber screen or not. This also means that, when used with the error snackbar modification described above, the "Try again" snackbar button continues to function even after navigating to another screen.

### How can we use this knowledge to improve the app?

Based on the above findings, we could potentially eliminate the loading screens that follow some user actions, allowing the user to continue to navigate freely through the app immediately after sending an API request. They would receive a success or failure snackbar on request completion, regardless of where in the app they ended up. The failure snackbar contains, if appropriate, a "Try again" button. This allows them to retry the request even if they have left the screen where they initiated it.

Note that input from design/UX would be required if we removed loading/saving/deleting indicators. I think these types of changes would need attention on a case by case basis. From a technical standpoint, moving our snackbar display callbacks from  `mutate` into `useMutation` would require changes at each call site. Code would likely have to be moved higher inside components in many cases so that it can be passed into the hook, which is defined near the top of the component. So while we'd want to take a global scope in designing changes to error handling, we couldn't just make a single code change in one location to implement the changes.

The changes described in the above sections would remove a class of possible silent failures, where navigating away from the original screen causes the failure snackbar to disappear, or never be shown at all.

**Comment by dumathane** [comment]
Closing and attaching to newly created Epic.
